### PR TITLE
Grafana configuration

### PR DESCRIPTION
**Pull request by atopuzov**
Provision data sources and dashboards declarativly
http://docs.grafana.org/administration/provisioning/

Example usage:
```
    provision = {
      enable = true;
      datasources = [
        {
          type = "prometheus";
          name = "prometheus";
          url = "http://localhost:9090";
        }
      ];
      dashboards = [
        {
          options.path = "/etc/nixos/grafana-dashboards";
        }
      ];
    };
```
This adds `prometheus` as datasource to grafana and a dashboard provider path "/etc/nixos/grafana-dashboards" which is scanned by grafana and dashboards are automatically added. 
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/modules/services/monitoring/grafana.nix [hunk 1]
```diff
@@ -50,6 +50,166 @@ let
...
+
+  # Get a submodule without any embedded metadata:
+  _filter = x: filterAttrs (k: v: k != "_module") x;
+
+  # http://docs.grafana.org/administration/provisioning/#datasources
+  grafanaTypes.datasourceConfig = types.submodule {
+    options = {
+      name = mkOption {
+        type = types.str;
+        default = "Grafana";
```
Comment (Ma27) [inline nixos/modules/services/monitoring/grafana.nix]: Naming a grafana datasource "Grafana" is not a good idea IMHO. As this field is required in the UI as well, I wouldn't set a default here.

File: nixos/modules/services/monitoring/grafana.nix [hunk 2]
```diff
@@ -50,6 +50,166 @@ let
...
+        description = "Name of the datasource. Required";
+      };
+      type = mkOption {
+        type = types.str;
```
Comment (Ma27) [inline nixos/modules/services/monitoring/grafana.nix]: As there's only a limited amount of possible data sources, we probably want to use something like `types.enum [ "graphite" "prometheus" /* and so on */ ]`

File: nixos/modules/services/monitoring/grafana.nix [hunk 3]
```diff
@@ -50,6 +50,166 @@ let
...
+      };
+      orgId = mkOption {
+        type = types.int;
+        default = 1;
+        description = "Org id. will default to orgId 1 if not specified";
+      };
+      url = mkOption {
+        type = types.nullOr types.str;
+        default = null;
+        description = "Url";
```
Comment (Ma27) [inline nixos/modules/services/monitoring/grafana.nix]: IIRC the datasource URL can't be empty, so I'd set it to `types.str` and don't provide a default. Also, the description could be improved.

File: nixos/modules/services/monitoring/grafana.nix [hunk 4]
```diff
@@ -50,6 +50,166 @@ let
...
+        description = "Basic auth username";
+      };
+      basicAuthPassword = mkOption {
+        type = types.nullOr types.str;
+        default = null;
+        description = "Basic auth password";
+      };
+      withCredentials = mkOption {
+        type = types.nullOr types.bool;
+        default = null;
```
Comment (Ma27) [inline nixos/modules/services/monitoring/grafana.nix]: Why `null` by default? This isn't and enabled by default and thus the default can be `false`.

File: nixos/modules/services/monitoring/grafana.nix [hunk 5]
```diff
@@ -359,6 +535,9 @@ in {
         ${optionalString (cfg.smtp.passwordFile != null) ''
           export GF_SMTP_PASSWORD="$(cat ${escapeShellArg cfg.smtp.passwordFile})"
         ''}
+        ${optionalString (cfg.provision.enable == true) ''
```
Comment (Ma27) [inline nixos/modules/services/monitoring/grafana.nix]: `optionalString cfg.provision.enable` should be sufficient.

File: nixos/modules/services/monitoring/grafana.nix [hunk 6]
```diff
@@ -50,6 +50,166 @@ let
...
+  dashboardConfiguration = {
+    apiVersion = 1;
+    providers = cfg.provision.dashboards;
+  };
+
+  dashboardFile = pkgs.writeText "dashboard.yaml" (builtins.toJSON dashboardConfiguration);
+
+  provisionConfDir = pkgs.stdenv.mkDerivation {
+    name = "grafana-provisioning";
+    buildCommand = ''
```
Comment (Ma27) [inline nixos/modules/services/monitoring/grafana.nix]: I guess that `runCommand` may be better here:

``` nix
pkgs.runCommand "grafana-provisioning" { } ''
  /* create provisioning derivation */
''
```

File: nixos/modules/services/monitoring/grafana.nix [hunk 7]
```diff
@@ -50,6 +50,166 @@ let
...
+      mkdir $out
+      mkdir $out/datasources
+      mkdir $out/dashboards
```
Comment (Ma27) [inline nixos/modules/services/monitoring/grafana.nix]: `mkdir -p $out/{datasources,dashboards}` may be easier.

File: nixos/modules/services/monitoring/grafana.nix [hunk 8]
```diff
@@ -50,6 +50,166 @@ let
...
+        default = "Grafana";
+        description = "Name of the datasource. Required";
+      };
+      type = mkOption {
+        type = types.str;
+        default = "graphite";
+        description = "Datasource type. Required";
+      };
+      access = mkOption {
+        type = types.str;
```
Comment (Ma27) [inline nixos/modules/services/monitoring/grafana.nix]: `types.enum []` may be a better type as well.

File: nixos/modules/services/monitoring/grafana.nix [hunk 9]
```diff
@@ -50,6 +50,166 @@ let
...
+        description = "Org id. will default to orgId 1 if not specified";
+      };
+      url = mkOption {
+        type = types.nullOr types.str;
+        default = null;
+        description = "Url";
+      };
+      password = mkOption {
+        type = types.nullOr types.str;
+        default = null;
```
Comment (Ma27) [inline nixos/modules/services/monitoring/grafana.nix]: please add an warning during the evaluation if this is set. The Nix store is world readable, so every user can see this.

Warnings can be added likes this:

``` nix
{
  /* your-module.nix */

  config = mkIf cfg.enable {
    warnings = optional (password != null) [ ''
       This is a warning.
    '' ];
  };
}
```

File: nixos/modules/services/monitoring/grafana.nix [hunk 10]
```diff
@@ -50,6 +50,166 @@ let
...
+      };
+      basicAuthUser = mkOption {
+        type = types.nullOr types.str;
+        default = null;
+        description = "Basic auth username";
+      };
+      basicAuthPassword = mkOption {
+        type = types.nullOr types.str;
+        default = null;
+        description = "Basic auth password";
```
Comment (Ma27) [inline nixos/modules/services/monitoring/grafana.nix]: the same issue mentioned for the db password applies here: please add a warning if this is set.

**Review by Ma27 — CHANGES_REQUESTED** [review]
Thanks a lot for doing this! I recently configured a Grafana and I missed such a feature! I tested this locally in a VM and it seems fine.

Below are some review comments. Feel free to ping me if you need further assistance.

Also, please rename your commits as described in the [contribution guidelines](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md), i.e. `nixos/grafana: implement dashboard & datasource provisioning`.

**Comment by atopuzov** [comment]
@Ma27 Updated per comments

**Review by Ma27 — APPROVED** [review]
Tested in a VM, I really like it :+1: 

I'll write some changelog, but I guess that this should be mergable then.

**Comment by Ma27** [comment]
rebased onto master and added a changelog entry.

**Comment by Ma27** [comment]
@atopuzov thanks!

**Comment by bhipple** [comment]
Thanks for the work here; declarative grafana provisioning looks really slick and I'm looking forward to trying it out in NixOS 19.03!

**Comment by arianvp** [comment]
I have found some bugs with this.  For some reason, non of the introduced options here actually get typechecked by the nixos module system.

E.g. you can omit any of the fields and it will just compile. Also if you mistype any of the fields it works, and it also works when you use the wrong type anywhere.

I have a feeling the typechekcing is somehow disabled due to the custom `apply` function used; but not sure.

e.g. the following config just works, whilst it's clearly wrong:

```
      grafana.provision.datasources = [
        {
          # name = "hello"; # name is mandatory but I can leave it out
          editable = "hello i should be a boolean but i made it a string";
          type = "prometheus";
          url = "http://127.0.0.1:9090"; # TODO make configurable
        }
      ];

```

**Comment by arianvp** [comment]
Nevermind I forgot to enable `provision.enable` ;)

**Pull request merged**